### DISCUSSION TITLE
Exporting CA with ca.tokenpassword=null

### DISCUSSION THREAD

**Discussion by donusry-privoro** [Keyfactor/ejbca-ce#39] [Q&A] [answered]
Hi everyone. I recently inherited an old development PKI and have been tasked to bring it in line with our production one. Long story short it involves moving the sub CAs to their own machines. Originally we thought exporting was a lost cause due to no one knowing the password, but as it turns out the install.properties has ca.tokenpassword set to null. The issue is when I try to export and give it no password, then I get the following:

`Export a token without password protection is not allowed.`

Is there something I can do to get around this?

EJBCA 6.5.0.5

Thank you.

**Comment by primetomas** [comment]
If ca.tokenpassword in install.properties was set to null, only affects the initially created Management CA, not any Sub CAs that you subsequently created. 
So the best quess is that you have either:
1. The default password in from cesecore.properties (see ca.keystorepass)
2. A password was defined when creating the SubCA, in which case this was set by someone who should have made a note of it

If "autoactivation" is enabled for the CA, the password is stored encrypted in the database and can be recovered. Let me know if the above does not fix it for you.

**Reply by donusry-privoro** [reply]
cesecore.properties file is not present. Every crypto token has auto-activation enabled. I do not see MariaDB or other DB running on the system, so I will assume it is using H2.

**Reply by mike-agrenius-kushner** [reply]
Check if it's in ejbca.properties instead, assuming that this might be an older installation which has been upgraded?

**Reply by donusry-privoro** [reply]
The only thing defined in there is appserver.home. This has not been updated since it was brought up back in 2017.

**Comment by primetomas** [comment]
Ok, your EJBCA is very old, so it may be slightly different. I hope you can adjust.
In CryptoTokenData in later versions of EJBCA, probably in CAData in your version(?), you have properties in the database.
> mysql -u ejbca -p ejbca
> select tokenProps from CryptoTokenData where tokenName='foobar';
You may have to look in CAData instead. If you are using H2 database, then find where the database files are stored, and you will be able to check the data in the fies directly with a text editor.

In tokenProps I have a base64 encoded field, you may have an field called "pin" in cleartext (don't remember from that version).

I do:
>cat > a.b64
>openssl base64 -d -in a.b64
And find: 
pin=6bc841b2745e2c95e042a68b4777b34c
Here the hex data is the "encrypted" password. You can feed this into the java method StringTools.pbeDecryptStringWithSha256Aes192 to decrypt it. In my case I modified StringToolsTest to decrypt it...using the setting in my config files for the encryption key (which should be unique).

Not trivial I know, but it's shouldn't be trivial out-of-the-box to decrypt this.

**Reply by donusry-privoro** [reply]
Thank you for your reply, though I am unsure how to continue. I found ejbcadb.h2.db in the home directory. Opening it in a text editor yields what looks like XML, and searching for things like "pin=" or "CAData" doesn't reveal anything spectacular. I tried opening the H2 database with H2 Console and I don't see any tables other than INFORMATION_SCHEMA.

**Reply by donusry-privoro** [reply]
Progress has been made. I was able to get a password following your instructions. However, I am now getting a new exception when trying to export. Originally it was always `Exception in thread "main" javax.ejb.EJBException: org.cesecore.keys.token.CryptoTokenAuthenticationFailedException: PKCS12 key store mac invalid - wrong password or corrupted file.`

Now I am getting `Exception in thread "main" javax.ejb.EJBException: org.bouncycastle.operator.OperatorCreationException: cannot create signer: Supplied key (org.bouncycastle.jcajce.provider.asymmetric.ec.BCECPrivateKey) is not a RSAPrivateKey instance`

I hope this means the password I extracted worked, and there is now some other (hopefully trivial) issue to work through. Looks to be related to JCE provider.

**Reply by mike-agrenius-kushner** [reply]
Oh, you're in luck – this one is easy. The CA has an EC key set as its encryption key – one of those things we should really just not allow. As you may know, EC can't be used out of the box for encryption, so the solution is just to change the encryption key to an RSA key.

**Reply by primetomas** [reply]
Can you share the full stacktrace from server.log, that will show more details.

**Reply by donusry-privoro** [reply]
@primetomas here is a tail of my server.log. [server.log](https://github.com/Keyfactor/ejbca-ce/files/8717147/server.log)

@kombatminipig I tried creating a new RSA 4096 keypair under the crypto token and I still wasn't able to export. Is there another encryption key that I need to change?

**Comment by primetomas** [comment, marked as answer]
Details how to fix the issue highlighted by @kombatminipig.
In the CA configuration you have a set of key aliases, one of them is called "keyEncryptKey". It can be either of two things:
- Either you have set the keyEncryptKey to an EC key, or you have "defaultKey" set to an EC key and no setting for "keyEncryptKey". 

The solution should be to set keyEncryptKey eplicitly to be an RSA key.

**Reply by donusry-privoro** [reply]
All of the key fields are set to signKey which is ECDSA. I created an RSA keypair under the crypto token, but don't see a way to change keyEncryptKey in either the WebUI or CLI.

<img width="333" alt="image" src="https://user-images.githubusercontent.com/104790463/169342522-85d5734f-d815-4d73-b9a6-8cd0f93fddb7.png">

**Reply by mike-agrenius-kushner** [reply]
Yeah, add it to the list of UI improvements we should make ^_^ Luckily there's a CLI command you can use, check out the man page for changecatoken:

bin/ejbca.sh ca changecatoken --help

It'll let you modify the keyEncryptKey

Cheers,
Mike

**Reply by donusry-privoro** [reply]
I do not have that command available. Here is what I have under CLI category CA. I dug through changecatokensignalg since it seemed to be the closest match. This is an old version of EJBCA, so it could be somewhere else?
<img width="1025" alt="image" src="https://user-images.githubusercontent.com/104790463/169346956-12aabd5f-d5cf-4748-bc6a-dd0b50289ae5.png">

**Reply by mike-agrenius-kushner** [reply]
Oh, yeah, you're at 6.5 😬 thought that command was ages old, but so is 6.5...

An upgrade to the latest community is pretty quick though? Just build it in the same environment and replace the EAR.

**Reply by donusry-privoro** [reply]
The upgrade guide isn't helping me. It just says "Upgrade to the latest version of EJBCA" for version 6.4+. I ran `ant upgrade` and `ant deployear`.

**Reply by mike-agrenius-kushner** [reply]
Uhm, odd question, but you've downloaded the latest CE and deployed it, right? If you don't want to check it out from here with GIT, you can find a tag at https://sourceforge.net/projects/ejbca/files/latest/download of 7.4.3.2 (We haven't gotten around to building tags here on github yet, but we will soon). 

As much as we'd love to, most PKI-systems are offline or very heavily firewalled, so there's no way for us to make an auto-update procedure. 

Built the zip from the above link, run deployear to build and deploy it into Wildfly and then you should be good to go. 

Cheers,
Mike

**Reply by donusry-privoro** [reply]
Not an odd question. No I did not, but now that you mention that it makes sense. This is what happens when you have an engineer unfamiliar with PKI doing sysadmin stuff :). I will work on upgrading. Thanks!

**Reply by mike-agrenius-kushner** [reply]
Ah, awesome =) Hopefully the upgrade guide makes a bit more sense from that perspective. Let me know, I'm personally invested at this point in you getting those keys exported.

Cheers,
Mike

**Reply by donusry-privoro** [reply]
I followed the upgrade guide, but I believe my JBoss instance (6.4) is too old. I spent the afternoon trying to get WildFly 18 working, but was unsuccessful.

This is the message from JBoss 6.4 when running standalone.sh:

`Failed to start service jboss.deployment.unit."ejbca.ear".INSTALL: org.jboss.msc.service.StartException in service jboss.deployment.unit."ejbca.ear".INSTALL: JBAS018733: Failed to process phase INSTALL of deployment "ejbca.ear"`

**Reply by mike-agrenius-kushner** [reply]
Not much I can do with just that error message I'm afraid, but we do have a pretty solid guide for appserver setup: https://doc.primekey.com/ejbca/ejbca-installation/application-servers/wildfly-18-jboss-eap-7-3

Have you followed it?

**Reply by donusry-privoro** [reply]
Yeah. That's what I used. I'm going to run through it again today. I might have messed up the database and datasource parts of the guide. I didn't see examples for H2, so I winged it.

**Reply by donusry-privoro** [reply]
I followed the guide to setup WildFly 18 except I manually copied in the db and ds config into standalone.xml from old JBoss. I then followed the EJBCA upgrade guide. When I run standalone.xml now I get exceptions related to incorrect password for the keystore and truststore.

[server.log](https://github.com/Keyfactor/ejbca-ce/files/8743246/server.log)

**Reply by donusry-privoro** [reply]
I was able to upgrade EJBCA enough to get a working command line, change the tokenprop, and export the CA. Thank you all so much for your help. Would have been nice to get the web interface working, but I will stop while I am ahead.

**Reply by mike-agrenius-kushner** [reply]
Then my work here is done. 

/me flies off into the distance.
